### PR TITLE
Change all active TEXT columns in MySQL app DB to LONGTEXT

### LINKED ISSUE #7006: MySQL doesn't support text columns > 64KB; Need to use LONGTEXT instead of TEXT
We have an SQL query that stops working when being saved and re-openend. The query itself is correct, can be run in a database client and is - at times - processed correctly by Metabase as well. The error happens in the following scenario:

1. Create a new question with the attached SQL statement. It can be executed/refreshed and the result data is displayed correctly.
2. Save the question.
3. Go back to the "Questions" section, search for the previously saved question and re-open it.
4. Metabase takes a while (presumably awaiting the result from the database) and then displays the following error message: "Data truncation: Data too long for column 'results' at row 1"
5. Using the "Refresh" button does not help either, the query does not retrieve any data.
6. Open the editor to modify the SQL statement. Add or remove a blank anywhere within the query.
7. When pusing the "Get Answer" button, data is now loaded and displayed successfully.
8. When saving the question and going back to step 3, the problem will repeat itself.

Additionally, we have observed the following:

- The problem does not occur for small result sets (it starts somewhere around 4,000 rows).
- Using the same statement in another question (stored under a different name) does not help.
- Altering the statement (to the extend possible) doesn't help either (we tried to replace the main SELECT by the sub-select directly behind the FROM clause).

Please find the simplified data model, the SQL query and the stacktrace attached.

[sql-statements.txt](https://github.com/metabase/metabase/files/1750037/sql-statements.txt)
[stacktrace.txt](https://github.com/metabase/metabase/files/1750021/stacktrace.txt)

==[ Environment ]===========================

- Your browser and the version: Microsoft Edge 41.16299.248.0
- Your operating system: Windows 10
- Your databases: MySQL
- Metabase version: 0.28.1
- Metabase hosting environment: Docker
- Metabase internal database: MySQL

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by jeff303**
Change all active TEXT columns in MySQL app DB to LONGTEXT

Add new text.type Liquibse property to dynamically select the correct text type to use, per DB (which is LONGTEXT for MySQL), very similar to the existing blob.type one

Adding new migrations that update all existing TEXT columns to LONGTEXT columns, for mariadb/mysql only

Update migrations linter to ensure no new text types get added, via a new predicate that searches for text types being added

### CODE REVIEW

File: resources/migrations/000_migrations.yaml [hunk 1]
```diff
@@ -8648,6 +8657,846 @@ databaseChangeLog:
                   constraints:
                     nullable: false
                   defaultValue: false
+  - changeSet:
+      id: 321
+      author: jeff303
+      comment: Added 0.42.0 - modify type of activity.details from text to ${text.type}
+        on mysql,mariadb
```
Comment (camsaul) [inline resources/migrations/000_migrations.yaml:8664]: is this supposed to be part of the comment? Maybe you should make it all one line or use one of the multiline markdown directives just to be safe

File: resources/migrations/000_migrations.yaml [hunk 2]
```diff
@@ -8648,6 +8657,846 @@ databaseChangeLog:
...
+      changes:
+        - modifyDataType:
+            tableName: activity
+            columnName: details
+            newDataType: ${text.type}
+      preConditions:
+        - onFail: MARK_RAN
+        - dbms:
+            type: mysql,mariadb
```
Comment (camsaul) [inline resources/migrations/000_migrations.yaml:8673]: I'm not 100% sure that you can list multiple types here, or if you need to use `or`... have you manually confirmed that this works for both MySQL and MariaDB? Maybe adding a test for MySQL/MariaDB that just confirms the application DB columns are the right types or something could be a good idea

**Review by camsaul — APPROVED** [review]
overall LGTM but I'd try to make all those comments one line to make sure nothing gets confused anywhere

File: resources/migrations/000_migrations.yaml (hunk 1 above)
Comment (jeff303) [reply resources/migrations/000_migrations.yaml:8664]: This is valid YAML (the "plain style" as described [here](https://stackoverflow.com/a/21699210)).  And this was even generated by our YAML library, using code from here: `https://www.notion.so/metabase/App-DB-Migrations-Tips-and-Tricks-b29b4a8048964ac0a0685b8e6ff19fa0`

But I can see about bulk updating the style to something else instead.

File: resources/migrations/000_migrations.yaml (hunk 2 above)
Comment (jeff303) [reply resources/migrations/000_migrations.yaml:8673]: The same thing is used in a couple other places in the file already, [ex](https://github.com/metabase/metabase/blame/a459f19ac8805614acb4709ae16e3aacafb2776a/resources/migrations/000_migrations.yaml#L5342-L5345).  But I can add a test (looks like `test-migrations` macro should work for this).

Comment (camsaul) [reply resources/migrations/000_migrations.yaml:8673]: You probably don't even need `test-migrations`, just write a test that checks the type of one of the app DB columns if we're running MySQL or MariaDB. You can use JDBC `DatabaseMetaData` to do that

**Comment by flamber** [comment]
We still need to create another changeset for 97 (query_cache.results), so existing instances gets converted to LONGBLOB

**Comment by camsaul** [comment]
I think we should do the QueryCache migration as a follow-on, mostly because we already have migrations 381 and 382 already so that one would have to be nonsequential anyway

File: resources/migrations/000_migrations.yaml (hunk 2 above)
Comment (jeff303) [reply resources/migrations/000_migrations.yaml]: I just wrote a comprehensive test for all app DBs.

**Comment by jeff303** [comment]
Opened #18815 to cover fixing `query_cache.results` on upgrade.

File: enterprise/backend/test/metabase_enterprise/serialization/cmd_test.clj [hunk 3]
```diff
@@ -22,7 +22,7 @@
   what we need."
   [& body]
   `(schema-migrations-test.impl/with-temp-empty-app-db [conn# :h2]
-     (schema-migrations-test.impl/run-migrations-in-range! conn# [0 999999]) ; this should catch all migrations)
+     (schema-migrations-test.impl/run-migrations-in-range! conn# [0 "v99.00-000"]) ; this should catch all migrations)
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/serialization/cmd_test.clj:25]: this should have been in the PR to switch to the new migration format but I missed this

File: bin/lint-migrations-file/src/change_set/common.clj [hunk 4]
```diff
@@ -15,7 +15,7 @@
 
 (s/def ::new-style-id
   (s/and string?
-         #(re-matches #"^v\d{2}\.\d{2}-\d{3}$" %)))
+         #(re-matches #"^v\d{2,}\.\d{2}-\d{3}$" %)))
```
Comment (camsaul) [inline bin/lint-migrations-file/src/change_set/common.clj:18]: this should have been in the PR to switch to the new migration format but I missed this. Minor tweak. Allow >2 digit major versions. So when we get to 100.0 we don't have things start breaking

**Pull request merged**